frame/cosmwasm/src/lib.rs

```diff
@@ -443,6 +445,40 @@ pub mod pallet {
 			let outcome = Self::do_extrinsic_execute(&mut shared, who, contract, funds, message);
 			Self::refund_gas(outcome, gas, shared.gas.remaining())
 		}
+
+		#[transactional]
+		#[pallet::weight(T::WeightInfo::migrate().saturating_add(*gas))]
+		pub fn migrate(
+			origin: OriginFor<T>,
+			contract: AccountIdOf<T>,
+			new_code_id: CosmwasmCodeId,
+			gas: u64,
+			message: ContractMessageOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let mut shared = Self::do_create_vm_shared(gas, InitialStorageMutability::ReadWrite);
+			let outcome =
+				Self::do_extrinsic_migrate(&mut shared, who, contract, new_code_id, message);
+			Self::refund_gas(outcome, gas, shared.gas.remaining())
+		}
+
+		#[transactional]
+		#[pallet::weight(T::WeightInfo::upload_and_instantiate(code.len(), funds.len()).saturating_add(*gas))]
+		pub fn upload_and_instantiate(
+			origin: OriginFor<T>,
+			code: ContractCodeOf<T>,
+			salt: ContractSaltOf<T>,
+			admin: Option<AccountIdOf<T>>,
+			label: ContractLabelOf<T>,
+			funds: FundsOf<T>,
+			gas: u64,
+			message: ContractMessageOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let next_code_id =
+				CurrentCodeId::<T>::get().checked_add(1).ok_or(Error::<T>::CodeIdOverflow)?;
```

Comment:
You can use the Nonce abstraction for this?

Reply:
We removed this extrinsic, instead we introduce a new identifier for code. Which is an `enum code_id | code_hash`. This way users won't need to wait for the code id to be generated and returned.